### PR TITLE
[charts] New, REST API

### PR DESCRIPTION
**Pull request by dpgaspar**
### CATEGORY

Choose one

- [ ] Bug Fix
- [X] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
New charts REST API. 

- Took the chance to make a more generic solution for `Model`s that implement a owners field and security.
- Improved field validation for owners and dashboards

image

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/views/api.py [hunk 1]
```diff
@@ -27,6 +27,7 @@
 from superset.utils import core as utils
 
 from .base import api, BaseSupersetView, handle_api_exception
+from .chart import api as chart_api  # pylint: disable=unused-import
 from .dashboard import api as dashboard_api  # pylint: disable=unused-import
 from .database import api as database_api  # pylint: disable=unused-import
```
Comment (villebro) [inline superset/views/api.py]: There was talk about absolute vs relative imports with @john-bodley some time back, and I believe there was consensus that we should always go for absolute paths when possible.

File: superset/views/chart/api.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        raise ValidationError("JSON not valid")
+
+
+def validate_owners(value):
+    owner = (
+        current_app.appbuilder.get_session.query(
+            current_app.appbuilder.sm.user_model.id
+        )
+        .filter_by(id=value)
+        .one_or_none()
```
Comment (villebro) [inline superset/views/chart/api.py]: I think a cleaner approach here is to use a try-catch block with `.one()`, as we don't expect there to be a `None` result.

File: superset/views/core.py [hunk 3]
```diff
@@ -105,8 +105,8 @@
     json_success,
     SupersetModelView,
 )
+from .chart import views as chart_views
 from .dashboard import views as dash_views
-from .dashboard.filters import DashboardFilter
 from .database import views as in_views
 from .utils import (
```
Comment (villebro) [inline superset/views/core.py]: Same here regarding absolute vs relative imports.

**Review by villebro — COMMENTED** [review]
A few comments on first pass

File: superset/views/core.py (hunk 3 above)
Comment (dpgaspar) [reply superset/views/core.py]: Will make a separate PR to address all relative imports on views

File: superset/views/dashboard/api.py [hunk 4]
```diff
@@ -80,32 +75,22 @@ def validate_slug_uniqueness(value):
 
 
 def validate_owners(value):
```
Comment (willbarrett) [inline superset/views/dashboard/api.py]: This method is duplicated in  superset/views/chart/api.py - can we extract it to a shared util?

**Review by willbarrett — COMMENTED** [review]
A couple of thoughts for you :)

File: superset/views/base.py [hunk 5]
```diff
@@ -520,6 +566,155 @@ def related(self, column_name: str, **kwargs):
...
+                id=item.data.id,
+            )
+        except SQLAlchemyError as e:
+            return self.response_422(message=str(e))
+
+    @expose("/<pk>", methods=["DELETE"])
+    @protect()
+    @check_ownership_and_item_exists
+    @safe
+    def delete(self, item):  # pylint: disable=arguments-differ
```
[the remark this answers is not in the document]
Comment (dpgaspar) [reply superset/views/base.py]: It is, delete will not process request data, just the id on the URI `/api/v1/<resource>/<id>`

File: superset/views/chart/api.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+
+def validate_json(value):
+    try:
+        utils.validate_json(value)
+    except SupersetException:
+        raise ValidationError("JSON not valid")
+
+
+def validate_owners(value):
```
[the remark this answers is not in the document]
Comment (dpgaspar) [reply superset/views/chart/api.py]: Makes sense

File: superset/views/dashboard/api.py (hunk 4 above)
Comment (dpgaspar) [reply superset/views/dashboard/api.py]: Yap, planning to DRY things a bit more, still a work in progress

File: superset/views/base.py [hunk 7]
```diff
@@ -396,6 +407,41 @@ def load(
         return super().load(data, many=many, partial=partial, **kwargs)
 
 
+class BaseOwnedSchema(BaseSupersetSchema):
```
[the remark this answers is not in the document]
Comment (dpgaspar) [reply superset/views/base.py]: done

File: superset/views/chart/api.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dpgaspar) [reply superset/views/chart/api.py]: Done

File: superset/views/dashboard/api.py (hunk 4 above)
Comment (dpgaspar) [reply superset/views/dashboard/api.py]: Done

File: superset/views/chart/api.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    edit_model_schema = ChartPutSchema()
+
+    order_rel_fields = {
+        "slices": ("slice_name", "asc"),
+        "owners": ("first_name", "asc"),
+    }
+    filter_rel_fields_field = {"owners": "first_name", "dashboards": "dashboard_title"}
+
+
+appbuilder.add_api(ChartRestApi)
```
Comment (craig-rueda) [inline superset/views/chart/api.py]: Move this into `app.py`

**Review by craig-rueda — CHANGES_REQUESTED** [review]
One nit, otherwise LGTM

**Review by craig-rueda — APPROVED** [review]

**Pull request merged**